pkg/httpwithcache/httpclient.go

```diff
@@ -0,0 +1,83 @@
+package httpwithcache
+
+import (
+	"database/sql"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+var cacheDB *sql.DB
+
+func SetDatabase(db *sql.DB) {
+	cacheDB = db
+}
+
+func Do(client *http.Client, req *http.Request) ([]byte, error) {
+	return DoWithCacheExpire(client, req, 2*time.Hour)
+}
+
+func DoWithCacheExpire(client *http.Client, req *http.Request, cacheExpire time.Duration) ([]byte, error) {
+	var cachedResponse []byte
+	var lastCached time.Time
+	var lastTried time.Time
+
+	endpoint := req.Method + " " + req.URL.String()
+	sqlerr := cacheDB.QueryRow(`SELECT response_body, created_at,
+	 tried_at FROM http_cache WHERE endpoint = $1`,
+		endpoint).Scan(&cachedResponse, &lastCached, &lastTried)
+	if sqlerr == nil {
+		if time.Since(lastCached) > cacheExpire && time.Since(lastTried) > cacheExpire {
+			reqWithoutContext, err := http.NewRequest(req.Method, req.URL.String(), req.Body)
+			if err == nil {
+				go updateCache(client, reqWithoutContext)
+			} else {
+				log.WithError(err).Errorf("Failed to create request")
+			}
+		}
+		return cachedResponse, nil
+	}
+
+	return updateCache(client, req)
+}
+
+func updateCache(client *http.Client, req *http.Request) ([]byte, error) {
+	endpoint := req.Method + " " + req.URL.String()
+	log.Printf("Update cache for %v", endpoint)
+	_, err := cacheDB.Exec(`INSERT INTO http_cache (endpoint, response_body, created_at, tried_at) 
+	VALUES ($1, $2, $3, $3) ON CONFLICT (endpoint) DO UPDATE SET tried_at = $3`, endpoint, "", time.Now().UTC())
```

Comment:
Yeah, this is possible, but very unlikely to happen, because:
After the first call to teamkatalogen, we will have the cached data in database, so the "insert" will always trigger a conflict after that. I.E. the response in database will never set to blank after it has the first value in database, which will probably happen when we test the code first time.
Setting the timestamp is to avoid starting duplicate go routines when a update is in progress. tried_at time stamp is an exclusive flag for update routines.

Reply:
> tried_at time stamp is an exclusive flag for update routines.
What do you mean by this? Is it a feature in Go?

Reply:
Ah, I missread the SQL query, now I understand.